components/service/glean/scripts/sdk_generator.gradle

```diff
@@ -112,7 +112,13 @@ ext.gleanGenerateMetricsAPI = {
     variant ->
 
     def sourceOutputDir =  "$buildDir/telemetry/src/${variant.name}/kotlin"
-    def fullNamespace = "${variant.applicationId}.GleanMetrics"
+    def applicationId = variant.applicationId
+    if (applicationId.endsWith(".debug")) {
```

Comment:
I'm afraid that applications can [manually define](https://developer.android.com/studio/build/application-id#change_the_application_id_for_build_variants) `applicationIdSuffix`(es) in their `build.gradle` file.

Luckily for us, turns out we have a way to get the  [`buildType`](https://android.googlesource.com/platform/tools/build/+/master/gradle/src/main/groovy/com/android/build/gradle/api/BaseVariant.java#85) from the `variant`. We can actually get the `applicationIdSuffix` from there. So this could become, more generally:

```
    def filteredAppId = variant.applicationId
    if (variant.buildType.applicationIdSuffix) {
        filteredAppId = filteredAppId.replaceAll(".$variant.buildType.applicationIdSuffix", "")
    }
```

To prevent future regressions, I would also add a suffix to the glean sample app, like [Fenix does](https://github.com/mozilla-mobile/fenix/blob/72b9d4ebdb3f9d9baeb849b3933efdd6eb915cf7/app/build.gradle#L39-L44).

Reply:
> We somehow need to make sure that the generated code is always in the same package - regardless of the application id.
> 
> This is a similar problem that the Android build faces when generating the `R` and `BuildConfig` classes. They are always in the same place (I assume from the `defaultConfig`). Maybe that's something we can try to do too (not sure how though!)?

This was the *killer* suggestion! I found:

```
    println("DEBUG: $variant.applicationId")
        Task buildConfig = variant.generateBuildConfig
        def test = buildConfig.getBuildConfigPackageName()
    println("DEBUG: $test")
```

Prints

> DEBUG: org.mozilla.samples.glean.LOL
> DEBUG: org.mozilla.samples.glean

The first is the flavoured id, the second is the package name used in the BuildConfig.java class.

Reply:
Cool.  Whatever we do here will also not need to break in the context of library-defined metrics.  Building the sample app should hopefully catch that, but we'll need to confirm...

Reply:
This is what I get in Fenix (wrapped in `android.applicationVariants.all { ... }`:

```
DEBUG: org.mozilla.fenix
-----
DEBUG: org.mozilla.fennec_aurora
DEBUG: org.mozilla.fenix
-----
DEBUG: org.mozilla.fenix.debug
DEBUG: org.mozilla.fenix
-----
DEBUG: org.mozilla.fenix
DEBUG: org.mozilla.fenix
-----
DEBUG: org.mozilla.firefox.debug
DEBUG: org.mozilla.fenix
-----
DEBUG: org.mozilla.firefox
DEBUG: org.mozilla.fenix
-----
DEBUG: org.mozilla.firefox_beta.debug
DEBUG: org.mozilla.fenix
-----
DEBUG: org.mozilla.firefox_beta
DEBUG: org.mozilla.fenix
-----
DEBUG: org.mozilla.fennec_aurora.debug
DEBUG: org.mozilla.fenix
-----
DEBUG: org.mozilla.fennec_aurora
DEBUG: org.mozilla.fenix
-----
DEBUG: org.mozilla.fenix.debug
DEBUG: org.mozilla.fenix
-----
DEBUG: org.mozilla.fenix
DEBUG: org.mozilla.fenix
-----
DEBUG: org.mozilla.firefox.debug
DEBUG: org.mozilla.fenix
-----
DEBUG: org.mozilla.firefox
DEBUG: org.mozilla.fenix
-----
DEBUG: org.mozilla.firefox_beta.debug
DEBUG: org.mozilla.fenix
-----
DEBUG: org.mozilla.firefox_beta
DEBUG: org.mozilla.fenix
-----
DEBUG: org.mozilla.fennec_aurora.debug
DEBUG: org.mozilla.fenix
-----
DEBUG: org.mozilla.fennec_aurora
DEBUG: org.mozilla.fenix
-----
DEBUG: org.mozilla.fenix.debug
DEBUG: org.mozilla.fenix
-----
DEBUG: org.mozilla.fenix
DEBUG: org.mozilla.fenix
-----
DEBUG: org.mozilla.firefox.debug
DEBUG: org.mozilla.fenix
-----
DEBUG: org.mozilla.firefox
DEBUG: org.mozilla.fenix
-----
DEBUG: org.mozilla.firefox_beta.debug
DEBUG: org.mozilla.fenix
-----
DEBUG: org.mozilla.firefox_beta
DEBUG: org.mozilla.fenix
```